service/history/stateBuilder.go

```diff
@@ -48,21 +50,24 @@ type (
 func newStateBuilder(shard ShardContext, msBuilder *mutableStateBuilder, logger bark.Logger) *stateBuilder {
 
 	return &stateBuilder{
-		shard:       shard,
-		msBuilder:   msBuilder,
-		domainCache: shard.GetDomainCache(),
-		logger:      logger,
+		shard:           shard,
+		clusterMetadata: shard.GetService().GetClusterMetadata(),
+		msBuilder:       msBuilder,
+		domainCache:     shard.GetDomainCache(),
+		logger:          logger,
 	}
 }
 
-func (b *stateBuilder) applyEvents(version int64, sourceClusterName string, domainID, requestID string,
-	execution shared.WorkflowExecution, history *shared.History, newRunHistory *shared.History) (*shared.HistoryEvent,
+func (b *stateBuilder) applyEvents(domainID, requestID string, execution shared.WorkflowExecution,
+	history *shared.History, newRunHistory *shared.History) (*shared.HistoryEvent,
 	*decisionInfo, *mutableStateBuilder, error) {
 	var lastEvent *shared.HistoryEvent
 	var lastDecision *decisionInfo
 	var newRunStateBuilder *mutableStateBuilder
 	for _, event := range history.Events {
 		lastEvent = event
+		// must set the current version, since this is standby here, not active
+		b.msBuilder.updateReplicationStateVersion(event.GetVersion())
```

Comment:
I'm not sure this is the right place to update CurrentStateVersion.  It is currently updated on workflow execution load.  I think we should just update the API on 'updateReplicationStateLastEventID' to also take in clusterName and lastWriteVersion and update both lastWriteVersion and lastWriteEventID as part of the same API.  Then we can call this explicitly from conflict resolver at the end like we update other mutableState fields.

Reply:
Let's hold onto this till Monday.  I think we should have a quick discussion before we land this.